wikidata/client.py

```diff
@@ -179,19 +180,28 @@ def decode_datavalue(self,
 
     def request(self, path: str) -> Union[
         bool, int, float, str,
-        Mapping[
-            str,
-            Union[bool, int, float, str, Mapping[str, object], Sequence]
-        ],
-        Sequence[Union[bool, int, float, str, Mapping[str, object], Sequence]]
+        Mapping[str, Union[bool, int, float, str,
+                           Mapping[str, object], Sequence]],
+        Sequence[Union[bool, int, float, str, Mapping[str, object], Sequence]],
+        None
     ]:
         logger = logging.getLogger(__name__ + '.Client.request')
         url = urllib.parse.urljoin(self.base_url, path)
         result = self.cache_policy.get(CacheKey(url))
         if result is None:
             logger.debug('%r: no cache; make a request...', url)
-            response = self.opener.open(url)
-            buffer_ = io.TextIOWrapper(response, encoding='utf-8')
+            try:
+                response = self.opener.open(url)
+            except urllib.error.HTTPError as e:
+                logger.debug('HTTP error code: %s', e.code, exc_info=True)
+                response = e.fp.read().decode('utf-8')  # type: ignore
+                if 'Invalid ID' in response:
```

Comment:
-  It would be better if it checks `e.code == 400`.

 -  Decoding `bytes` to `str` seems unnecessary.  `bytes` also supports `in` operator.

 -  [According to the docs][1]:

    > Though being an exception (a subclass of [`URLError`][2]), an [`HTTPError`][1] can also function as a non-exceptional file-like return value (the same thing that [`urlopen()`][3] returns).

    There you can write like `e.read()` instead of `e.fp.read()`.

Long story short:

```suggestion
                if e.code == 400 and b'Invalid ID' in e.read():
```

[1]: https://docs.python.org/3/library/urllib.error.html#urllib.error.HTTPError
[2]: https://docs.python.org/3/library/urllib.error.html#urllib.error.URLError
[3]: https://docs.python.org/3/library/urllib.request.html#urllib.request.urlopen